Environment Variable Expansion

adqm opened the pull request:
This changeset:
* implements the changes discussed at the bottom of #539, including updating documentation
* makes aliases do path expansion at well (as discussed in #338)

I think it is all working, and I hope I didn't break anything else.

Also, separately, I'm not quite sure how the Travis and Appveyor autobuild things work, but I am trying here to make them run in quiet mode to avoid printing a dot for each of the test cases (there are really a lot of them now, after the last PR was merged, and printing the dots really seems to slow things down).

adqm commented:
Woops this broke things on Windows, it seems.  I'll ping when fixed.

scopatz commented:
Thanks @adqm. Did a quick review and it seemed good. Didn't see the win error. Just ping.

adqm commented:
Okay, @scopatz, ready for another look, I think.  Changes from last time:
* included a modified version of `expandvars` that uses xonsh's environment instead of `os.environ` to do replacement in strings
* fixed the regexes that check for which ensurer to use, to avoid `DIRSTACK_SIZE` being interpreted as a `\w*DIRS` variable
* skipped several test cases on Windows, where spoofed environments were necessary
* made the environment variable lookups always fall back to the default values so that just checking, e.g., the value of `$XONSH_STORE_STDOUT` works now (shows the default, rather than raising an IndexError).  Is there a reason things didn't work this way before?

scopatz commented on xonsh/tools.py:
```diff
@@ -909,3 +909,50 @@ def check_for_partial_string(x):
+        i, j = m.span(0)
+        name = m.group(1)
+        if name.startswith(start) and name.endswith(end):
+            name = eval(name[1:-1]) # change for non-POSIX ${...}
```
`ast.literal_eval()` maybe? Or is this a full eval? Also do you need the current locals and globals?  Seems like an empty execution context is desirable here.

adqm replied:
I think this has to be a full eval, but I think it should happen in the current locals and globals.

My thought is that if we have `x = "HOME"`, both of the following should work:
```
>>> ${x}
>>> echo "${x}"
```

literal eval wouldn't handle that case.

scopatz replied:
Understood. Got it. That will be super cool to have.

scopatz replied:
But I think you then want to use the xonsh eval or at least the `__xonsh_ctx__` as the execution context.  As written this will use the function's globals and locals.

adqm replied:
Yes, sorry; that was what I meant by the comment above; we will want to eval in the xonsh context.

scopatz commented on xonsh/tools.py:
```diff
@@ -909,3 +909,50 @@ def check_for_partial_string(x):
+_varprog = None
+
+
+def expandvars(path):
```
This seems way more complex than using [`Env.replacve_env()`](http://xonsh.org/api/environ.html#xonsh.environ.Env.replace_env) to replaces the contents of os.environ with detyped env, and then call os.expandvars().  Is this doing something more xonsh-specific that I am not seeing? Or is it more performant in some way?

adqm replied:
Well, one thing is that I think it is nice if this works with the `${..}` syntax (none of the `expandvars` in `os` will do that).

More important is the fallback to default values, which wouldn't work with `replace_env`.

scopatz replied:
Ahhh OK these are both good enough for me :).

scopatz commented on xonsh/built_ins.py:
```diff
@@ -180,7 +182,9 @@ def expand_path(s):
         ENV.replace_env()
-    return os.path.expanduser(os.path.expandvars(s))
+    if ENV.get('EXPAND_ENV_VARS'):
+        s = expandvars(s)
```
This still calls `ENV.repleace_env()

scopatz replied:
This still calls `ENV.replace_env()`, which now seems either unnecessary or totally sufficient.

adqm replied:
Are `replace_env` and `undo_replace_env` used someplace else I am not seeing?  I am planning on just deleting them unless they are useful someplace else.

scopatz commented on xonsh/tools.py:
```diff
@@ -909,3 +909,50 @@ def check_for_partial_string(x):
+    """Expand shell variables of form $var and ${var}.  Unknown variables
+    are left unchanged."""
+    global _varprog
+    ENV = builtins.__xonsh_env__
```
Should this be lower-case env?

scopatz commented on xonsh/tools.py:
```diff
@@ -909,3 +909,50 @@ def check_for_partial_string(x):
+    ENV = builtins.__xonsh_env__
+    if isinstance(path, bytes):
+        path = path.decode(encoding=env.get('XONSH_ENCODING'),
+                           errors=env.get('XONSH_ENCODING_ERRORS'))
```
Or should this be upper case env?

adqm replied:
you're right that one of them should change :)

scopatz commented on xonsh/tools.py:
```diff
@@ -909,3 +909,50 @@ def check_for_partial_string(x):
+        return path
+    if not _varprog:
+        import re
+        _varprog = re.compile(r'\$(\w+|\{[^}]*\})', re.ASCII)
```
This doesn't handle `%name%` completion on windows, which someone will probably complain about :) https://docs.python.org/3.5/library/os.path.html#os.path.expandvars  Shouldn't be too hard to modify.

adqm replied:
Ah, yes.  That's what I get for copying from `posixpath.py` and not even looking at the windows equivalent.  I can work that in.

scopatz commented:
> made the environment variable lookups always fall back to the default values so that just checking, e.g., the value of $XONSH_STORE_STDOUT works now (shows the default, rather than raising an IndexError). Is there a reason things didn't work this way before?

Nope that was an oversight!  Thanks for fixing.

scopatz commented:
Thanks for all of the hard work here!

adqm commented:
Alright, ready for another look.  This time I:
* modified `expandvars` to do the `%name%` expansion on Windows
* removed `replace_env` and `undo_replace_env`
* modified the eval to evaluate in the right context

scopatz commented on xonsh/environ.py:
```diff
@@ -242,24 +243,6 @@ def detype(self):
-        if self._orig_env is not None:
-            os.environ.clear()
-            os.environ.update(self._orig_env)
-            self._orig_env = None
```
I think that these are still potentially useful to users.  Do you mind putting them back in?

scopatz commented:
Looks great to me, other than the method removal. Basically ready to merge.

scopatz commented:
I can put them back in if you'd like too.

adqm commented:
I just put them back in, so we should be good to go, I think (assuming the tests pass).  Let me know if you have any other thoughts/suggestions.

I'll also take a look at your history PR tonight.

scopatz commented:
Awesome!
